File: app/src/main/kotlin/tech/mmarca/openvitals/features/manualentry/activity/ActivityEntryState.kt
package tech.mmarca.openvitals.features.manualentry.activity

import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.material.icons.outlined.SentimentVeryDissatisfied
import androidx.compose.material.icons.outlined.SentimentDissatisfied
import androidx.compose.material.icons.outlined.SentimentSatisfied
import androidx.compose.material.icons.outlined.SentimentVerySatisfied
import androidx.compose.material.icons.Icons
import tech.mmarca.openvitals.features.manualentry.*
import tech.mmarca.openvitals.features.manualentry.activity.*
import tech.mmarca.openvitals.features.manualentry.activity.recording.*
import tech.mmarca.openvitals.features.manualentry.activity.routeimport.*
import tech.mmarca.openvitals.features.manualentry.body.*
import tech.mmarca.openvitals.features.manualentry.hydration.*
import tech.mmarca.openvitals.features.manualentry.mindfulness.*
import tech.mmarca.openvitals.features.manualentry.vitals.*



import androidx.compose.runtime.Immutable
import java.time.Instant
import tech.mmarca.openvitals.core.presentation.ScreenError
import tech.mmarca.openvitals.core.presentation.toScreenError
import tech.mmarca.openvitals.domain.model.ActivityPauseInterval
import tech.mmarca.openvitals.domain.model.ActivityRecordingMarker
import tech.mmarca.openvitals.domain.model.BleRecordingSampleBuffer
import tech.mmarca.openvitals.domain.model.CoMapsNavigationSnapshot
import tech.mmarca.openvitals.domain.model.ExerciseLapData
import tech.mmarca.openvitals.domain.model.ExerciseRoutePoint
import tech.mmarca.openvitals.domain.model.HeartRateSample
import tech.mmarca.openvitals.domain.model.PlannedExerciseData
import tech.mmarca.openvitals.R

enum class ActivityEntryError {
    INVALID_VALUE,
    MISSING_WRITE_PERMISSION,
    ROUTE_IMPORT_FAILED,
    LOCATION_PERMISSION_NEEDED,
    NOTIFICATION_PERMISSION_NEEDED,
    ACTIVITY_RECOGNITION_PERMISSION_NEEDED,
    RECORDING_FAILED,
    WRITE_FAILED,
    PLAN_NOT_FOUND,
}

enum class ActivityEntryMode {
    START_HUB,
    MANUAL,
    ROUTE_IMPORT,
    RECORDING,
}

enum class ActivityEntryField {
    ACTIVITY_TYPE,
    TITLE,
    START_DATE,
    START_TIME,
    DURATION,
    REPETITIONS,
    DISTANCE,
    ELEVATION,
    ACTIVE_CALORIES,
    TOTAL_CALORIES,
}

enum class ActivityEntryValidationError(
    val field: ActivityEntryField,
) {
    ACTIVITY_TYPE_DOES_NOT_SUPPORT_ROUTE(ActivityEntryField.ACTIVITY_TYPE),
    START_DATE_INVALID(ActivityEntryField.START_DATE),
    START_TIME_INVALID(ActivityEntryField.START_TIME),
    START_TIME_AFTER_ROUTE_START(ActivityEntryField.START_TIME),
    DURATION_INVALID(ActivityEntryField.DURATION),
    REPETITIONS_INVALID(ActivityEntryField.REPETITIONS),
    DISTANCE_INVALID(ActivityEntryField.DISTANCE),
    DISTANCE_UNSUPPORTED(ActivityEntryField.DISTANCE),
    ELEVATION_INVALID(ActivityEntryField.ELEVATION),
    ELEVATION_UNSUPPORTED(ActivityEntryField.ELEVATION),
    ACTIVE_CALORIES_INVALID(ActivityEntryField.ACTIVE_CALORIES),
    TOTAL_CALORIES_INVALID(ActivityEntryField.TOTAL_CALORIES),
    TOTAL_CALORIES_BELOW_ACTIVE(ActivityEntryField.TOTAL_CALORIES),
}

enum class ActivityRepetitionEntryMode {
    TOTAL,
    SETS,
}

/** A plan about to be run step by step: resolved type and the flattened steps. */
data class ActivityGuidedPlan(
    val plan: PlannedExerciseData,
    val activityType: ActivityEntryType,
    val steps: List<ActivityPlanRunStep>,
)

/** The plan a session is logged against; the title is null when only the id is known. */
data class ActivityLinkedPlan(
    val id: String,
    val title: String?,
)

/**
 * How the session felt, on a four-step scale. Material glyphs, not emoji:
 * emoji ignore the colour scheme and cannot carry selection state.
 */
enum class ActivityEntryFeeling(
    val icon: ImageVector,
    val labelRes: Int,
    val noteText: String,
) {
    GREAT(
        icon = Icons.Outlined.SentimentVerySatisfied,
        labelRes = R.string.activity_entry_feeling_great,
        noteText = "Felt great.",
    ),
    GOOD(
        icon = Icons.Outlined.SentimentSatisfied,
        labelRes = R.string.activity_entry_feeling_good,
        noteText = "Felt good.",
    ),
    HARD(
        icon = Icons.Outlined.SentimentDissatisfied,
        labelRes = R.string.activity_entry_feeling_hard,
        noteText = "Felt hard.",
    ),
    ROUGH(
        icon = Icons.Outlined.SentimentVeryDissatisfied,
        labelRes = R.string.activity_entry_feeling_rough,
        noteText = "Felt rough.",
    ),
}

/** One set. A plan can put a different exercise or a timed hold in the sequence. */
data class ActivityRepetitionSetInput(
    /** The rep count, or the seconds when [isDuration]. */
    val repetitionsText: String = "",
    val restMinutesText: String = "",
    /** Segment type for this set when it differs from the activity type's own; null means the activity's. */
    val segmentType: Int? = null,
    /** A stored step label ("Push-ups"); null falls back to the segment type's name. */
    val label: String? = null,
    val isDuration: Boolean = false,
)

@Immutable
data class ActivityEntryUiState(
    val mode: ActivityEntryMode = ActivityEntryMode.START_HUB,
    val activityTypes: List<ActivityEntryType> = DefaultActivityEntryTypes,
    val selectedActivityType: ActivityEntryType = DefaultActivityEntryTypes.first(),
    val titleText: String = "",
    val selectedFeeling: ActivityEntryFeeling? = null,
    val notesText: String = "",
    val startDateText: String = "",
    val startTimeText: String = "",
    val durationMinutesText: String = "30",
    val distanceText: String = "",
    val elevationText: String = "",
    val activeCaloriesText: String = "",
    val totalCaloriesText: String = "",
    val repetitionMode: ActivityRepetitionEntryMode = ActivityRepetitionEntryMode.TOTAL,
    val repetitionTotalText: String = "",
    val repetitionSets: List<ActivityRepetitionSetInput> = listOf(ActivityRepetitionSetInput()),
    val linkedPlan: ActivityLinkedPlan? = null,
    /** Set while the recording setup is showing a plan to run; cleared with the rest of the form. */
    val guidedPlan: ActivityGuidedPlan? = null,
    /** Uncompleted plans for today and later, by start time; the start hub's list. */
    val hubPlans: List<PlannedExerciseData> = emptyList(),
    /** Recently completed plans, newest first, one per title: what "Repeat" offers. */
    val recentPlans: List<PlannedExerciseData> = emptyList(),
    val isLoadingHubPlans: Boolean = false,
    val hubPlansError: ScreenError? = null,
    /** False when this device's Health Connect has no planned-exercise feature at all. */
    val hubPlansAvailable: Boolean = true,
    val isSavingAsPlan: Boolean = false,
    /** One-shot request for the screen to open the builder on this plan. */
    val pendingBuilderPlanId: String? = null,
    val importedRoute: RouteFileImport? = null,
    val recordedPauseIntervals: List<ActivityPauseInterval> = emptyList(),
    val recordedLaps: List<ExerciseLapData> = emptyList(),
    val recordedMarkers: List<ActivityRecordingMarker> = emptyList(),
    val writePermissions: Set<String> = emptySet(),
    val canWrite: Boolean = false,
    val isCheckingPermission: Boolean = true,
    val isImportingRoute: Boolean = false,
    val isSavingEntry: Boolean = false,
    val entryError: ActivityEntryError? = null,
    val detailError: ScreenError? = null,
    val validationErrors: Set<ActivityEntryValidationError> = emptySet(),
    val editRecordId: String? = null,
    val isRecordingDraft: Boolean = false,
    val saveCompleted: Boolean = false,
    val recordedBleSamples: BleRecordingSampleBuffer = BleRecordingSampleBuffer(),
    /** CoMaps guidance banked during the recording. App-local only. */
    val recordedCoMapsSamples: List<CoMapsNavigationSnapshot> = emptyList(),
    val sessionHeartRateSamples: List<HeartRateSample> = emptyList(),
    /** When the effort stopped in a recovery test, written as a trailing rest segment. */
    val recordedRecoveryStartTime: Instant? = null,
) {
    val routePoints: List<ExerciseRoutePoint>
        get() = importedRoute?.points.orEmpty()

    val isEditMode: Boolean
        get() = editRecordId != null
}

File: lib/features/manualentry/activity/activity_entry_state.dart
import 'package:collection/collection.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../../core/presentation/command_state.dart';
import '../../../core/presentation/screen_error.dart';
import '../../../domain/model/activity_models.dart';
import '../../../domain/model/ble_sensor_models.dart';
import '../../../domain/model/heart_models.dart';
import '../../../domain/model/activity_entry_types.dart';
import 'routeimport/route_file_parser.dart';

part 'activity_entry_state.freezed.dart';

/// Port of the Kotlin `ActivityEntryState.kt` — the manual-entry / recording
/// screen UI state plus its supporting enums and value types.

enum ActivityEntryError {
  invalidValue,
  missingWritePermission,
  routeImportFailed,
  locationPermissionNeeded,
  notificationPermissionNeeded,
  activityRecognitionPermissionNeeded,
  recordingFailed,
  writeFailed,
}

/// The screen's high-level form mode (Kotlin `ActivityEntryMode`; renamed to
/// avoid clashing with the navigation-intent `ActivityEntryMode`).
enum ActivityEntryFormMode {
  chooseSource,
  planActivityPicker,
  planPicker,
  manual,
  routeImport,
  recording,
}

enum ActivityEntryField {
  activityType,
  title,
  startDate,
  startTime,
  duration,
  repetitions,
  distance,
  elevation,
  activeCalories,
  totalCalories,
}

enum ActivityEntryValidationError {
  activityTypeDoesNotSupportRoute(ActivityEntryField.activityType),
  trainingPlanTitleRequired(ActivityEntryField.title),
  startDateInvalid(ActivityEntryField.startDate),
  startTimeInvalid(ActivityEntryField.startTime),
  startTimeAfterRouteStart(ActivityEntryField.startTime),
  durationInvalid(ActivityEntryField.duration),
  repetitionsInvalid(ActivityEntryField.repetitions),
  distanceInvalid(ActivityEntryField.distance),
  distanceUnsupported(ActivityEntryField.distance),
  elevationInvalid(ActivityEntryField.elevation),
  elevationUnsupported(ActivityEntryField.elevation),
  activeCaloriesInvalid(ActivityEntryField.activeCalories),
  totalCaloriesInvalid(ActivityEntryField.totalCalories),
  totalCaloriesBelowActive(ActivityEntryField.totalCalories);

  const ActivityEntryValidationError(this.field);

  final ActivityEntryField field;
}

enum ActivityRepetitionEntryMode { total, sets }

enum ActivityEntryFeeling {
  great('😀', 'Great', 'Felt great.'),
  good('🙂', 'Good', 'Felt good.'),
  hard('😓', 'Hard', 'Felt hard.'),
  rough('😖', 'Rough', 'Felt rough.');

  const ActivityEntryFeeling(this.emoji, this.label, this.noteText);

  final String emoji;
  final String label;
  final String noteText;
}

/// Port of the Kotlin `ActivityRepetitionSetInput`.
class ActivityRepetitionSetInput {
  const ActivityRepetitionSetInput({
    this.repetitionsText = '',
    this.restMinutesText = '',
  });

  final String repetitionsText;
  final String restMinutesText;

  ActivityRepetitionSetInput copyWith({
    String? repetitionsText,
    String? restMinutesText,
  }) =>
      ActivityRepetitionSetInput(
        repetitionsText: repetitionsText ?? this.repetitionsText,
        restMinutesText: restMinutesText ?? this.restMinutesText,
      );

  @override
  bool operator ==(Object other) =>
      other is ActivityRepetitionSetInput &&
      other.repetitionsText == repetitionsText &&
      other.restMinutesText == restMinutesText;

  @override
  int get hashCode => Object.hash(repetitionsText, restMinutesText);
}

/// Port of the Kotlin `ActivityPlannedWorkoutBaseline`.
class ActivityPlannedWorkoutBaseline {
  const ActivityPlannedWorkoutBaseline({
    required this.planId,
    required this.activityTypeId,
    required this.titleText,
    required this.notesText,
    required this.startDateText,
    required this.startTimeText,
    required this.durationMinutesText,
    required this.repetitionMode,
    required this.repetitionTotalText,
    required this.repetitionSets,
  });

  final String planId;
  final String activityTypeId;
  final String titleText;
  final String notesText;
  final String startDateText;
  final String startTimeText;
  final String durationMinutesText;
  final ActivityRepetitionEntryMode repetitionMode;
  final String repetitionTotalText;
  final List<ActivityRepetitionSetInput> repetitionSets;

  @override
  bool operator ==(Object other) =>
      other is ActivityPlannedWorkoutBaseline &&
      other.planId == planId &&
      other.activityTypeId == activityTypeId &&
      other.titleText == titleText &&
      other.notesText == notesText &&
      other.startDateText == startDateText &&
      other.startTimeText == startTimeText &&
      other.durationMinutesText == durationMinutesText &&
      other.repetitionMode == repetitionMode &&
      other.repetitionTotalText == repetitionTotalText &&
      const ListEquality<ActivityRepetitionSetInput>()
          .equals(other.repetitionSets, repetitionSets);

  @override
  int get hashCode => Object.hash(
        planId,
        activityTypeId,
        titleText,
        notesText,
        startDateText,
        startTimeText,
        durationMinutesText,
        repetitionMode,
        repetitionTotalText,
        Object.hashAll(repetitionSets),
      );
}

@freezed
abstract class ActivityEntryUiState with _$ActivityEntryUiState {
  const ActivityEntryUiState._();

  const factory ActivityEntryUiState({
    @Default(ActivityEntryFormMode.chooseSource) ActivityEntryFormMode mode,
    required ActivityEntryType selectedActivityType,
    @Default('') String titleText,
    ActivityEntryFeeling? selectedFeeling,
    @Default('') String notesText,
    @Default('') String startDateText,
    @Default('') String startTimeText,
    @Default('30') String durationMinutesText,
    @Default('') String distanceText,
    @Default('') String elevationText,
    @Default('') String activeCaloriesText,
    @Default('') String totalCaloriesText,
    @Default(ActivityRepetitionEntryMode.total)
    ActivityRepetitionEntryMode repetitionMode,
    @Default('') String repetitionTotalText,
    @Default([ActivityRepetitionSetInput()])
    List<ActivityRepetitionSetInput> repetitionSets,
    @Default(<PlannedExerciseData>[]) List<PlannedExerciseData> plannedWorkouts,
    String? selectedPlannedWorkoutId,
    ActivityPlannedWorkoutBaseline? selectedPlannedWorkoutBaseline,
    String? selectedPlannedWorkoutActivityTypeId,
    @Default(false) bool isLoadingPlannedWorkouts,
    @Default(false) bool isSavingPlannedWorkout,
    RouteFileImport? importedRoute,
    @Default(<ActivityPauseInterval>[])
    List<ActivityPauseInterval> recordedPauseIntervals,
    @Default(<ExerciseLapData>[]) List<ExerciseLapData> recordedLaps,
    @Default(<ActivityRecordingMarker>[])
    List<ActivityRecordingMarker> recordedMarkers,
    @Default(<String>{}) Set<String> writePermissions,
    @Default(false) bool canWrite,
    @Default(true) bool isCheckingPermission,

    /// The route/FIT file import: its own command, because it is a second
    /// failable action on this form and it fails for its own reasons.
    @Default(CommandState<void>.idle()) CommandState<void> routeImport,

    /// Writing (or updating) the activity. [CommandSuccess] is consumed exactly
    /// once by the screen, which then leaves the route.
    @Default(CommandState<void>.idle()) CommandState<void> save,
    ActivityEntryError? entryError,

    /// The detail behind a non-command [entryError] — a permission probe, a
    /// planned-workout read, a recording that would not start. A save or an
    /// import that failed carries its own [ScreenError] in its command; read
    /// them together through [blockingError].
    ScreenError? detailError,
    @Default(<ActivityEntryValidationError>{})
    Set<ActivityEntryValidationError> validationErrors,
    String? editRecordId,
    @Default(false) bool isRecordingDraft,
    @Default(BleRecordingSampleBuffer()) BleRecordingSampleBuffer recordedBleSamples,
    @Default(<HeartRateSample>[]) List<HeartRateSample> sessionHeartRateSamples,
  }) = _ActivityEntryUiState;

  List<ActivityEntryType> get activityTypes => defaultActivityEntryTypes;

  bool get isSavingEntry => save is CommandRunning<void>;

  bool get isImportingRoute => routeImport is CommandRunning<void>;

  /// The error line the form renders under [entryError]: whichever of the two
  /// commands failed, else the detail of the last non-command failure.
  ScreenError? get blockingError => switch (save) {
        CommandFailure<void>(:final error) => error,
        _ => switch (routeImport) {
            CommandFailure<void>(:final error) => error,
            _ => detailError,
          },
      };

  List<ExerciseRoutePoint> get routePoints =>
      importedRoute?.points ?? const [];

  bool get isEditMode => editRecordId != null;

  bool get hasSelectedPlannedWorkoutChanges {
    final planId = selectedPlannedWorkoutId;
    if (planId == null) return false;
    final baseline = selectedPlannedWorkoutBaseline;
    if (baseline == null || baseline.planId != planId) return false;
    return plannedWorkoutBaseline(planId) != baseline;
  }

  /// Port of the Kotlin `ActivityEntryUiState.activitySaveNotes`.
  String? activitySaveNotes() {
    final feelingText = selectedFeeling?.noteText;
    final trimmedNote = notesText.trim();
    final noteText = trimmedNote.isEmpty ? null : trimmedNote;
    final parts = <String>[
      ?feelingText,
      ?noteText,
    ];
    final joined = parts.join('\n\n');
    return joined.isEmpty ? null : joined;
  }

  /// Port of the Kotlin `ActivityEntryUiState.plannedWorkoutBaseline`.
  ActivityPlannedWorkoutBaseline plannedWorkoutBaseline(String planId) =>
      ActivityPlannedWorkoutBaseline(
        planId: planId,
        activityTypeId: selectedActivityType.id,
        titleText: titleText.trim(),
        notesText: activitySaveNotes() ?? '',
        startDateText: startDateText.trim(),
        startTimeText: startTimeText.trim(),
        durationMinutesText: durationMinutesText.trim(),
        repetitionMode: repetitionMode,
        repetitionTotalText: repetitionTotalText.trim(),
        repetitionSets: [
          for (final set in repetitionSets)
            set.copyWith(
              repetitionsText: set.repetitionsText.trim(),
              restMinutesText: set.restMinutesText.trim(),
            ),
        ],
      );
}
